<a href="https://colab.research.google.com/github/Nirdesh-arc/Nirdesh_Khadka_Concept_and_Tech_of_AI/blob/main/Week5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np

#Define the cost function
def cost_function(X, Y, W):
  """ Parameters:
  This function finds the Mean Square Error.
  Input parameters:
  X: Feature Matrix
  Y: Target Matrix
  W: Weight Matrix
  Output Parameters:
  cost: accumulated mean square error."""
  m=len(Y)
  y_pred = np.sum((X.dot(W)-Y)**2)/(2*m)
  return y_pred
#Test the cost function
if __name__ == "__main__":
  X_test = np.array([[1, 2], [3, 4], [5, 6]])
  Y_test = np.array([3, 7, 11])
  W_test = np.array([1, 1])
  cost = cost_function(X_test, Y_test, W_test)
  if cost == 0:
    print("Proceed Further")
  else:
    print("something went wrong: Reimplement a cost function")
    print("Cost function output:", cost_function(X_test, Y_test, W_test))



Proceed Further


In [ ]:
def gradient_descent(X, Y, W, alpha, iterations):
  """
  Perform gradient descent to optimize the parameters of a linear regression model.
  Parameters:
  X (numpy.ndarray): Feature matrix (m x n).
  Y (numpy.ndarray): Target vector (m x 1).
  W (numpy.ndarray): Initial guess for parameters (n x 1).
  alpha (float): Learning rate.
  iterations (int): Number of iterations for gradient descent.
  Returns:
  tuple: A tuple containing the final optimized parameters (W_update) and the history of cost values
  .
  W_update (numpy.ndarray): Updated parameters (n x 1).
  cost_history (list): History of cost values over iterations.
  """
  # Initialize cost history
  cost_history = [0] * iterations
  # Number of samples
  m = len(Y)
  for iteration in range(iterations):
  # Step 1: Hypothesis Values
    Y_pred = X.dot(W)
  # Step 2: Difference between Hypothesis and Actual Y
    loss = Y_pred - Y
  # Step 3: Gradient Calculation
    dw = (X.T.dot(loss))/(m)
  # Step 4: Updating Values of W using Gradient
    W = W - alpha * dw
  # Step 5: New Cost Value
    cost_history.append(cost_function(X, Y, W))
    cost_history[iteration] = cost
  return W, cost_history
  # Generate random test data
  np.random.seed(0) # For reproducibility
  X = np.random.rand(100, 3) # 100 samples, 3 features
  Y = np.random.rand(100)
  W = np.random.rand(3) # Initial guess for parameters
  # Set hyperparameters
  alpha = 0.01
  iterations = 1000
  # Test the gradient_descent function
  final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)
  # Print the final parameters and cost history
  print("Final Parameters:", final_params)
  print("Cost History:", cost_history)

In [ ]:
# Model Evaluation - RMSE
def rmse(Y, Y_pred):
  """
  This Function calculates the Root Mean Squres.
  Input Arguments:
  Y: Array of actual(Target) Dependent Varaibles.
  Y_pred: Array of predeicted Dependent Varaibles.
  Output Arguments:
  rmse: Root Mean Square.
  """
  m = len(Y)
  rmse = np.sqrt((1/m)*(np.sum((Y - Y_pred) ** 2)))
  return rmse

In [ ]:
# Model Evaluation - R2
def r2(Y, Y_pred):
  """
  This Function calculates the R Squared Error.Input Arguments:
  Y: Array of actual(Target) Dependent Varaibles.
  Y_pred: Array of predeicted Dependent Varaibles.
  Output Arguments:
  rsquared: R Squared Error.
  """
  mean_y = np.mean(Y)
  ss_tot = sum((Y- mean_y)**2)
  ss_res = sum((Y- Y_pred)**2)
  r2 = 1 - (ss_res/ss_tot)
  return r2

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
def main():
  # Step 1: Load the dataset
  data = pd.read_csv('student.csv')
  # Step 2: Split the data into features (X) and target (Y)
  X = data[['Math', 'Reading']].values # Features: Math and Reading marks
  Y = data['Writing'].values # Target: Writing marks
  # Step 3: Split the data into training and test sets (80% train, 20% test)
  X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)
  # Step 4: Initialize weights (W) to zeros, learning rate and number of iterations
  W = np.zeros(X_train.shape[1]) # Initialize weights
  alpha = 0.000001 # Learning rate
  iterations = 1000 # Number of iterations for gradient descent
  # Step 5: Perform Gradient Descent
  W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)
  # Step 6: Make predictions on the test set
  Y_pred = np.dot(X_test, W_optimal)
  # Step 7: Evaluate the model using RMSE and R-Squared
  model_rmse = rmse(Y_test, Y_pred)
  model_r2 = r2(Y_test, Y_pred)
  # Step 8: Output the results
  print("Final Weights:", W_optimal)
  print("Cost History (First 10 iterations):", cost_history[:10])
  print("RMSE on Test Set:", model_rmse)
  print("R-Squared on Test Set:", model_r2)
  # Execute the main function

if __name__ == "__main__":
  main()


Final Weights: [0.47308856 0.52582   ]
Cost History (First 10 iterations): [np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0)]
RMSE on Test Set: 5.856694748793876
R-Squared on Test Set: 0.8629707528684534


To-Do 11 – Present Your Findings:

Model Evaluation:
I assessed whether my model overfits underfits, or performs acceptably. Overall, the model demonstrates acceptable performance.with a Test RMSE of 5.8 and a Test R² of 0.86. The R² score of 0.86 indicates that the model effectively captures the relationship between Math and Reading scores and Writing scores. The RMSE of 5.86 is reasonable given the scale of the scores.

However, the cost history shows all zeros for the first 10 iterations. This could be due to:

The learning rate (0.00001) being too small to show noticeable cost reduction early,

Initial weights being accidentally close to optimal, or

A potential issue in the gradient calculation.

Learning Rate Experiments:
I experimented with various learning rates to observe their impact on convergence:

α = 0.000001: Too small, resulting in very slow convergence.

α = 0.00001: Current rate; provides good R² but convergence is not fully evident.

α = 0.0001: Likely optimal; faster convergence with similar accuracy.

α = 0.001 – 0.01: Too high; risk of divergence or overshooting during training.